### ISSUE TITLE
Uninitialized constant GraphQL::Metrics

### ISSUE THREAD

**Issue by andersklenke** [Shopify/graphql-metrics#16]
Hi there

I was trying to test this library in a Rails 6 project. Unfortunately I can't make it work. When I try to use it I get a `uninitialized constant GraphQL::Metrics` error.

Gemfile
```
...
gem "graphql"
gem "graphql-batch"
gem "graphql-metrics", require: "graphql_metrics"
...
```

api_schema.rb
```rb
require "simple_analyzer"

class ApiSchema < GraphQL::Schema
  query(Types::QueryType)
  mutation(Types::MutationType)

  use GraphQL::Execution::Interpreter
  use GraphQL::Analysis::AST
  use GraphQL::Execution::Errors

  instrument :query, GraphQL::Metrics::Instrumentation.new
  query_analyzer SimpleAnalyzer
  tracer GraphQL::Metrics::Tracer.new

  use GraphQL::Pagination::Connections
  use(GraphQL::Tracing::NewRelicTracing, set_transaction_name: true)
  use GraphQL::Batch

  cursor_encoder(URLSafeBase64Encoder)
```

simple_analyzer.rb
```rb
class SimpleAnalyzer < GraphQL::Metrics::Analyzer
  ANALYZER_NAMESPACE = :capture_all_metrics_analyzer_namespace

  def initialize(query_or_multiplex)
    super

    # `query` is defined on instances of objects inheriting from GraphQL::Metrics::Analyzer
    ns = query.context.namespace(ANALYZER_NAMESPACE)
    ns[:simple_extractor_results] = {}
  end

    # @param metrics [Hash] Query metrics, including a few details about the query document itself, as well as runtime
    # timings metrics, intended to be compatible with the Apollo Tracing spec:
    # https://github.com/apollographql/apollo-tracing#response-format
    #
    # {
    #   operation_type: "query",
    #   operation_name: "PostDetails",
    #   query_start_time: 1573833076.027327,
    #   query_duration: 2.0207119999686256,
    #   parsing_start_time_offset: 0.0010339999571442604,
    #   parsing_duration: 0.0008190000080503523,
    #   validation_start_time_offset: 0.0030819999519735575,
    #   validation_duration: 0.01704599999357015,
    # }
    #
    # You can use these metrics to track high-level query performance, along with any other details you wish to
    # manually capture from `query` and/or `query.context`.
  def query_extracted(metrics)
    custom_metrics_from_context = {
      request_id: query.context[:request_id],
      # ...
    }

      # You can make use of captured metrics here (logging to Kafka, request logging etc.)
      # log_metrics(:fields, metrics)
      #
      # Or store them on the query context:
    store_metrics(:queries, metrics.merge(custom_metrics_from_context))
  end

    # For use after controller:
    # class GraphQLController < ActionController::Base
    #   def graphql_query
    #     query_result = graphql_query.result.to_h
    #     do_something_with_metrics(query.context[:simple_extractor_results])
    #     render json: graphql_query.result
    #   end
    # end

    # @param metrics [Hash] Field selection metrics, including resolver timings metrics, also adhering to the Apollo
    # Tracing spec referred to above.
    #
    # `resolver_timings` is populated any time a field is resolved (which may be many times, if the field is nested
    # within a list field e.g. a Relay connection field).
    #
    # `lazy_resolver_timings` is only populated by fields that are resolved lazily (for example using the
    # graphql-batch gem) or that are otherwise resolved with a Promise. Any time spent in the field's resolver to
    # prepare work to be done "later" in a Promise, or batch loader will be captured in `resolver_timings`. The time
    # spent actually doing lazy field loading, including time spent within a batch loader can be obtained from
    # `lazy_resolver_timings`.
    #
    # {
    #   field_name: "id",
    #   return_type_name: "ID",
    #   parent_type_name: "Post",
    #   deprecated: false,
    #   path: ["post", "id"],
    #   resolver_timings: [
    #     start_time_offset: 0.011901999998372048,
    #     duration: 5.999987479299307e-06
    #   ],
    #   lazy_resolver_timings: [
    #     start_time_offset: 0.031901999998372048,
    #     duration: 5.999987479299307e-06
    #   ],
    # }
  def field_extracted(metrics)
    store_metrics(:fields, metrics)
  end

  # @param metrics [Hash] Argument usage metrics, including a few details about the query document itself, as well
  # as resolver timings metrics, also ahering to the Apollo Tracing spec referred to above.
  # {
  #   argument_name: "id",
  #   argument_type_name: "ID",
  #   parent_field_name: "post",
  #   parent_field_type_name: "QueryRoot",
  #   default_used: false,
  #   value_is_null: false,
  #   value: <GraphQL::Query::Arguments::ArgumentValue>,
  # }
  #
  # `value` is exposed here, in case you want to get access to the argument's definition, including the type
  # class which defines it, e.g. `metrics[:value].definition.metadata[:type_class]`
  def argument_extracted(metrics)
    store_metrics(:arguments, metrics)
  end

  private

  def store_metrics(context_key, metrics)
    ns = query.context.namespace(ANALYZER_NAMESPACE)
    ns[:simple_extractor_results][context_key] ||= []
    ns[:simple_extractor_results][context_key] << metrics
  end
end
```

Let me know if you need more info 😊

Thanks

**Comment by swalkinshaw** [comment]
Can you try `gem 'graphql-metrics', 'require: 'graphql/metrics'` ? Though it should work without the `require` so if it doesn't we've probably done something wrong.

**Comment by andersklenke** [comment]
@swalkinshaw with `gem "graphql-metrics", require: "graphql/metrics"`
I get the following error when I try to start the Rails server
```
Traceback (most recent call last):
bin/rails: Bootsnap::LoadPathCache::FallbackScan
  38: from bin/rails:4:in `<main>'
  37: from /Users/a/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.1/lib/active_support/dependencies.rb:325:in `require'
  36: from /Users/a/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.1/lib/active_support/dependencies.rb:291:in `load_dependency'
  35: from /Users/a/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.1/lib/active_support/dependencies.rb:325:in `block in require'
  34: from /Users/a/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.6/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:31:in `require'
  33: from /Users/a/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.6/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:22:in `require_with_bootsnap_lfi'
  32: from /Users/a/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.6/lib/bootsnap/load_path_cache/loaded_features_index.rb:92:in `register'
  31: from /Users/a/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.6/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:23:in `block in require_with_bootsnap_lfi'
  30: from /Users/a/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.6/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:23:in `require'
  29: from /Users/a/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/railties-6.0.1/lib/rails/commands.rb:18:in `<main>'
  28: from /Users/a/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/railties-6.0.1/lib/rails/command.rb:46:in `invoke'
  27: from /Users/a/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/railties-6.0.1/lib/rails/command/base.rb:65:in `perform'
  26: from /Users/a/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/thor-0.20.3/lib/thor.rb:387:in `dispatch'
  25: from /Users/a/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/thor-0.20.3/lib/thor/invocation.rb:126:in `invoke_command'
  24: from /Users/a/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/thor-0.20.3/lib/thor/command.rb:27:in `run'
  23: from /Users/a/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/railties-6.0.1/lib/rails/commands/server/server_command.rb:138:in `perform'
  22: from /Users/a/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/railties-6.0.1/lib/rails/commands/server/server_command.rb:138:in `tap'
  21: from /Users/a/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/railties-6.0.1/lib/rails/commands/server/server_command.rb:141:in `block in perform'
  20: from /Users/a/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.1/lib/active_support/dependencies.rb:325:in `require'
  19: from /Users/a/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.1/lib/active_support/dependencies.rb:291:in `load_dependency'
  18: from /Users/a/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/activesupport-6.0.1/lib/active_support/dependencies.rb:325:in `block in require'
  17: from /Users/a/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.6/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:31:in `require'
  16: from /Users/a/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.6/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:22:in `require_with_bootsnap_lfi'
  15: from /Users/a/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.6/lib/bootsnap/load_path_cache/loaded_features_index.rb:92:in `register'
  14: from /Users/a/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.6/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:23:in `block in require_with_bootsnap_lfi'
  13: from /Users/a/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.6/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:23:in `require'
  12: from /Users/a/code/tonsser/tonsser-api/config/application.rb:36:in `<main>'
  11: from /Users/a/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/bundler-2.1.4/lib/bundler.rb:174:in `require'
  10: from /Users/a/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/bundler-2.1.4/lib/bundler/runtime.rb:58:in `require'
   9: from /Users/a/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/bundler-2.1.4/lib/bundler/runtime.rb:58:in `each'
   8: from /Users/a/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/bundler-2.1.4/lib/bundler/runtime.rb:69:in `block in require'
   7: from /Users/a/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/bundler-2.1.4/lib/bundler/runtime.rb:69:in `each'
   6: from /Users/a/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/bundler-2.1.4/lib/bundler/runtime.rb:74:in `block (2 levels) in require'
   5: from /Users/a/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.6/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:27:in `require'
   4: from /Users/a/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.6/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:41:in `rescue in require'
   3: from /Users/a/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.6/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:22:in `require_with_bootsnap_lfi'
   2: from /Users/a/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.6/lib/bootsnap/load_path_cache/loaded_features_index.rb:89:in `register'
   1: from /Users/a/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.6/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:23:in `block in require_with_bootsnap_lfi'
/Users/a/.rbenv/versions/2.6.5/lib/ruby/gems/2.6.0/gems/bootsnap-1.4.6/lib/bootsnap/load_path_cache/core_ext/kernel_require.rb:23:in `require': cannot load such file -- graphql/metrics (LoadError)
``` 

If I don't require anything I get the same error as before `uninitialized constant GraphQL::Metrics`, when trying to run a query against GraphQL.

**Comment by chrisbutcher** [comment]
@andersklenke Apologies for the confusion! The README has been updated to document version 3.0.0 which exists in the maser branch but has not yet been released to Rubygems. 

I'll ping again here when the new gem version is released. At that point, the instructions will make sense.

**Comment by swalkinshaw** [comment]
You can also do this in the meantime:

```ruby
gem 'graphql-metrics', '~> 3.0.0', github: 'Shopify/graphql-metrics'
```

**Comment by andersklenke** [comment]
Thanks @chrisbutcher 👌🏼

**Comment by chrisbutcher** [comment]
https://rubygems.org/gems/graphql-metrics/versions/3.0.0 released!

Thanks for reporting the issue, @andersklenke !

**Comment by andersklenke** [comment]
@chrisbutcher No problem.

Sorry to bother again, but when I tried:
Gemfile
```
gem "graphql"
gem "graphql-batch"
gem "graphql-metrics", "~> 3.0"
```

I get this when running `bundle install`
``` 
Bundler could not find compatible versions for gem "graphql":
  In snapshot (Gemfile.lock):
    graphql (= 1.10.4)

  In Gemfile:
    graphql

    graphql-batch was resolved to 0.4.2, which depends on
      graphql (>= 1.3, < 2)

    graphql-metrics (~> 3.0) was resolved to 3.0.0, which depends on
      graphql (~> 1.9.15)

Running `bundle update` will rebuild your snapshot from scratch, using only
the gems in your Gemfile, which may resolve the conflict.
``` 

Can you change the requirement of graphql `~> 1.9.15` to something less specific, please? 😊

**Comment by chrisbutcher** [comment]
Apologies for that. I'll fix ASAP.

I'm curious, what version of graphql-ruby are you using? AST Analyzers which this gem now uses only came in with version ~1.9 of that gem.

On a related note, removing that constraint and bumping graphql-ruby to 1.10.4 errors out with in graphql-metrics 3.0.0 tests 👀

**Comment by andersklenke** [comment]
We're on the following versions
```
graphql (1.10.4)
graphql-batch (0.4.2)
```

**Comment by chrisbutcher** [comment]
@andersklenke graphql-metrics 4.0.0 is now released, with graphql-ruby 1.10.8+ support

FYI graphql-ruby 1.10.0 to 1.10.7 versions are incompatible with graphql-metrics, due to a regression in argument metadata availability that was fixed in https://github.com/rmosolgo/graphql-ruby/pull/2881

**Comment by andersklenke** [comment]
@chrisbutcher Thanks! Great work 💪🏼
